compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs

```diff
@@ -399,3 +419,90 @@ checkDFlags Options {..} dflags@DynFlags {..}
             \ or the dependency."
+
+-- Expand SDK package dependencies using the SDK root path.
+-- E.g. `daml-trigger` --> `$DAML_SDK/daml-libs/daml-trigger.dar`
+-- When invoked outside of the SDK, we will only error out
+-- if there is actually an SDK package so that
+-- When there is no SDK
+expandSdkPackages :: [FilePath] -> IO [FilePath]
+expandSdkPackages dars = do
+    mbSdkPath <- handleIO (\_ -> pure Nothing) $ Just <$> getSdkPath
+    mapM (expand mbSdkPath) dars
+  where
+    isSdkPackage fp = takeExtension fp `notElem` [".dar", ".dalf"]
+    expand mbSdkPath fp
+      | fp `elem` basePackages = pure fp
+      | isSdkPackage fp = case mbSdkPath of
+            Just sdkPath -> pure $ sdkPath </> "daml-libs" </> fp <.> "dar"
+            Nothing -> fail $ "Cannot resolve SDK dependency '" ++ fp ++ "'. Use daml assistant."
+      | otherwise = pure fp
+
+
+mkPackageFlag :: UnitId -> PackageFlag
+mkPackageFlag unitId = ExposePackage "--package " (UnitIdArg unitId) (ModRenaming True [])
+
+mkBaseUnits :: Maybe String -> [UnitId]
+mkBaseUnits optMbPackageName
+  | optMbPackageName == Just "daml-prim" =
+      []
+  | optMbPackageName == Just damlStdlib =
+      [ stringToUnitId "daml-prim" ]
+  | otherwise =
+      [ stringToUnitId "daml-prim"
+      , stringToUnitId damlStdlib ]
+
+dependantUnitsFromDamlYaml :: Maybe FilePath -> IO [UnitId]
+dependantUnitsFromDamlYaml mbRoot = do
+  let dir = fromMaybe "." mbRoot
+  withCurrentDirectory dir $ do
+    (deps,dataDeps) <- depsFromDamlYaml (ProjectPath dir)
+    deps <- expandSdkPackages (filter (`notElem` basePackages) deps)
+    calcUnitsFromDeps (deps ++ dataDeps)
+
+depsFromDamlYaml :: ProjectPath -> IO ([FilePath],[FilePath])
+depsFromDamlYaml projectPath = do
+  try (readProjectConfig projectPath) >>= \case
+    Left (_::ConfigError) -> return ([],[])
+    Right project -> return $ projectDeps project
+
+projectDeps :: ProjectConfig -> ([FilePath],[FilePath])
+projectDeps project = do
+  let deps = fromMaybe [] $ either (error . show) id $ queryProjectConfig ["dependencies"] project
+  let dataDeps = fromMaybe [] $ either (error . show) id $ queryProjectConfig ["data-dependencies"] project
+  (deps,dataDeps)
+
+calcUnitsFromDeps :: [FilePath] -> IO [UnitId]
+calcUnitsFromDeps deps = do
+  let (fpDars, fpDalfs) = partition ((== ".dar") . takeExtension) deps
+  entries <- mapM mainEntryOfDar fpDars
+  let dalfsFromDars =
+        [ ( dropExtension $ takeFileName $ ZipArchive.eRelativePath e
+          , BSL.toStrict $ ZipArchive.fromEntry e)
+        | e <- entries ]
+  dalfsFromFps <-
+    forM fpDalfs $ \fp -> do
+      bs <- BS.readFile fp
+      pure (dropExtension $ takeFileName fp, bs)
+  let mainDalfs = dalfsFromDars ++ dalfsFromFps
+  flip mapMaybeM mainDalfs $ \(name, dalf) -> runMaybeT $ do
+    (pkgId, _package) <-
+        liftIO $
+        either (fail . DA.Pretty.renderPretty) pure $
+        Archive.decodeArchive Archive.DecodeAsMain dalf
```

Comment:
```suggestion
        Archive.decodeArchivePackageId dalf
```
See https://github.com/digital-asset/daml/pull/4550

Reply:
rebased, and made this change